File: src/connectors/claude.py
"""
Claude API Connector
Analyze lifelogs using Anthropic Claude
"""

import logging
import json
from typing import Dict, List, Any
from anthropic import Anthropic


class ClaudeConnector:
    """Wrapper for Anthropic Claude API"""

    def __init__(self, api_key: str, model: str = "claude-sonnet-4-5-20250929"):
        """
        Initialize Claude connector

        Args:
            api_key: Anthropic API key
            model: Claude model to use
        """
        self.api_key = api_key
        self.model = model
        self.logger = logging.getLogger("nexus.claude")

        self.client = Anthropic(api_key=api_key)

    async def analyze_priorities(
        self,
        lifelogs: List[Dict[str, Any]],
        period: str = "today"
    ) -> Dict[str, List[Dict[str, Any]]]:
        """
        Analyze lifelogs and extract priorities

        Args:
            lifelogs: List of lifelog objects with transcripts
            period: Time period (today/week)

        Returns:
            Dictionary with engagements, demandes, deadlines
        """
        try:
            # Prepare transcript text
            transcripts = self._format_lifelogs(lifelogs)

            if not transcripts:
                self.logger.warning("No transcripts to analyze")
                return {"engagements": [], "demandes": [], "deadlines": []}

            # Build prompt
            prompt = self._build_analysis_prompt(transcripts, period)

            self.logger.info(f"Analyzing {len(lifelogs)} lifelogs with Claude")

            # Call Claude API
            message = self.client.messages.create(
                model=self.model,
                max_tokens=4096,
                messages=[{
                    "role": "user",
                    "content": prompt
                }]
            )

            # Parse response
            response_text = message.content[0].text
            priorities = self._parse_response(response_text)

            self.logger.info(
                f"Extracted {len(priorities.get('engagements', []))} engagements, "
                f"{len(priorities.get('demandes', []))} demandes, "
                f"{len(priorities.get('deadlines', []))} deadlines"
            )

            return priorities

        except Exception as e:
            self.logger.error(f"Failed to analyze priorities: {e}")
            return {"engagements": [], "demandes": [], "deadlines": []}

    def _format_lifelogs(self, lifelogs: List[Dict[str, Any]]) -> str:
        """Format lifelogs for analysis"""
        formatted = []

        for log in lifelogs:
            date = log.get("date", "Unknown date")
            title = log.get("title", "Untitled")
            transcript = log.get("transcript", "")

            if transcript:
                formatted.append(f"""
--- Lifelog: {title} ({date}) ---
{transcript}
""")

        return "\n".join(formatted)

    def _build_analysis_prompt(self, transcripts: str, period: str) -> str:
        """Build prompt for priority detection"""
        return f"""Tu es un assistant IA spécialisé dans l'analyse de conversations et notes vocales pour Christian Boulet, fractional CTO.

Ta mission : Analyser les transcripts ci-dessous et identifier les priorités d'action.

Critères de détection :

1. **ENGAGEMENTS PRIS** - Actions que Christian a promis de faire
   - Phrases comme "je vais...", "je te reviens avec...", "je m'engage à..."
   - Promesses faites à des clients/prospects
   - Actions spécifiques mentionnées

2. **DEMANDES REÇUES** - Requêtes nécessitant une action de Christian
   - Questions directes : "Peux-tu...", "J'aurais besoin de...", "Pourrais-tu..."
   - Requêtes clients à traiter
   - Informations demandées

3. **DEADLINES** - Urgences temporelles
   - Dates explicites mentionnées
   - "Avant [date]", "Pour [jour]", "D'ici [deadline]"
   - Échéances importantes

Période analysée : {period}

Transcripts :
{transcripts}

Retourne un objet JSON structuré comme ceci (et UNIQUEMENT du JSON valide, rien d'autre) :

{{
  "engagements": [
    {{
      "title": "Description courte de l'engagement",
      "description": "Détails supplémentaires",
      "confidence": 0.95,
      "source": "Conversation avec [nom] - [date]"
    }}
  ],
  "demandes": [
    {{
      "title": "Description de la demande",
      "description": "Contexte et détails",
      "confidence": 0.90,
      "source": "Conversation avec [nom] - [date]"
    }}
  ],
  "deadlines": [
    {{
      "title": "Action avec deadline",
      "description": "Détails",
      "date": "YYYY-MM-DD",
      "confidence": 0.85,
      "source": "Conversation avec [nom] - [date]"
    }}
  ]
}}

IMPORTANT : Retourne UNIQUEMENT le JSON, pas de texte avant ou après."""

    def _parse_response(self, response: str) -> Dict[str, List[Dict[str, Any]]]:
        """Parse Claude's JSON response"""
        try:
            # Extract JSON from response (in case there's surrounding text)
            start = response.find('{')
            end = response.rfind('}') + 1

            if start >= 0 and end > start:
                json_str = response[start:end]
                priorities = json.loads(json_str)
                return priorities
            else:
                self.logger.error("No JSON found in Claude response")
                return {"engagements": [], "demandes": [], "deadlines": []}

        except json.JSONDecodeError as e:
            self.logger.error(f"Failed to parse Claude response as JSON: {e}")
            self.logger.debug(f"Response was: {response}")
            return {"engagements": [], "demandes": [], "deadlines": []}

    def is_connected(self) -> bool:
        """Check if connector is configured"""
        return bool(self.api_key)

File: supabase/functions/_shared/clients/claude.ts
/**
 * Claude (Anthropic) API Client
 * AI-powered priority detection and analysis
 */

import Anthropic from 'npm:@anthropic-ai/sdk@^0.20.0';
import type {
  LimitlessConversation,
  Priority,
  ClaudeAnalysisResponse,
} from '../types/index.ts';

export class ClaudeClient {
  private client: Anthropic;
  private model: string;

  constructor(apiKey: string, model: string = 'claude-sonnet-4-5-20250929') {
    this.client = new Anthropic({ apiKey });
    this.model = model;
  }

  /**
   * Analyze conversations and extract priorities
   * @param conversations - Array of Limitless conversations
   * @param context - Optional additional context
   * @returns Detected priorities with confidence scores
   */
  async analyzePriorities(
    conversations: LimitlessConversation[],
    context?: string
  ): Promise<ClaudeAnalysisResponse> {
    const startTime = Date.now();

    try {
      console.log(`[Claude] Analyzing ${conversations.length} conversations...`);

      // Build conversation context
      const conversationText = this.buildConversationContext(conversations);

      // Construct analysis prompt
      const prompt = this.buildAnalysisPrompt(conversationText, context);

      // Call Claude API
      const response = await this.client.messages.create({
        model: this.model,
        max_tokens: 4096,
        temperature: 0.3, // Lower temperature for consistent extraction
        messages: [
          {
            role: 'user',
            content: prompt,
          },
        ],
      });

      // Parse Claude's response
      const analysisText = response.content[0].type === 'text'
        ? response.content[0].text
        : '';

      const priorities = this.parsePrioritiesFromResponse(analysisText, conversations);

      const processingTime = Date.now() - startTime;

      console.log(`[Claude] Detected ${priorities.length} priorities in ${processingTime}ms`);

      return {
        priorities,
        summary: analysisText,
        confidence: this.calculateOverallConfidence(priorities),
        processingTime,
      };

    } catch (error) {
      console.error('[Claude] Analysis failed:', error);
      throw error;
    }
  }

  /**
   * Build conversation context for Claude
   */
  private buildConversationContext(conversations: LimitlessConversation[]): string {
    return conversations
      .map((conv, index) => {
        const date = new Date(conv.startTime).toLocaleDateString('fr-CA');
        const time = new Date(conv.startTime).toLocaleTimeString('fr-CA');

        return `
--- Conversation ${index + 1} ---
Date: ${date} à ${time}
${conv.title ? `Sujet: ${conv.title}` : ''}
${conv.participants ? `Participants: ${conv.participants.join(', ')}` : ''}

Transcript:
${conv.transcript}
${conv.summary ? `\nRésumé: ${conv.summary}` : ''}
`;
      })
      .join('\n\n');
  }

  /**
   * Build analysis prompt for Claude
   */
  private buildAnalysisPrompt(conversationText: string, context?: string): string {
    return `Tu es NEXUS, l'assistant IA de Christian Boulet, fractional CTO chez Boulet Stratégies TI.

Ta mission: analyser les conversations de Christian et extraire TOUTES les priorités potentielles.

${context ? `\nContexte additionnel:\n${context}\n` : ''}

# Critères de détection (INCLUSIFS)

Cherche ces 3 types de priorités:

1. **ENGAGEMENTS** - Christian a promis ou dit qu'il va faire quelque chose
   - Exemples EXPLICITES: "je vais te revenir avec une proposition", "je t'envoie mon CV"
   - Exemples IMPLICITES: "je vais le régler", "il faut que je...", "je dois..."
   - Même vague, si ça indique une intention d'action future, INCLURE
   - Contexte souvent: @ordi (travail ordinateur)

2. **DEMANDES** - Quelqu'un a demandé quelque chose à Christian OU Christian a identifié une tâche à faire
   - Exemples: "peux-tu m'envoyer...", "j'aurais besoin que...", "il faudrait que..."
   - INCLURE aussi les auto-demandes: "il faut que je vérifie", "je devrais appeler"
   - Contexte: @appels (si retour de call), @ordi (si travail)

3. **DEADLINES** - Échéances mentionnées (même vagues)
   - Exemples PRÉCIS: "avant vendredi", "deadline 30 octobre"
   - Exemples VAGUES: "demain", "la semaine prochaine", "bientôt"
   - Accepter les deadlines implicites aussi

# Consignes PERMISSIVES

- ✅ MIEUX VAUT inclure une priorité douteuse que de la manquer
- ✅ ACCEPTER les engagements vagues ou informels ("je vais le régler", "faut que je...")
- ✅ ACCEPTER les intentions même sans deadline précise
- ✅ Score de confiance bas (0.3-0.5) pour les priorités vagues, élevé (0.7-1.0) pour les explicites
- ✅ Estimer la durée réaliste (5min, 15min, 30min, 1h, 2h+)
- ✅ Assigner un contexte (@appels, @ordi, @agenda, @attente)

# Format de réponse

Retourne un JSON structuré:

\`\`\`json
{
  "priorities": [
    {
      "title": "Titre court et actionnable",
      "type": "engagement|demande|deadline",
      "description": "Description détaillée avec contexte",
      "context": "@appels|@ordi|@agenda|@attente|@courses",
      "duration": "5min|15min|30min|1h|2h+",
      "confidence": 0.95,
      "source": "Conversation avec [Nom]",
      "dueDate": "2025-01-30" // SEULEMENT pour deadlines
    }
  ]
}
\`\`\`

# Conversations à analyser

${conversationText}

# Ton analyse (JSON uniquement)`;
  }

  /**
   * Parse priorities from Claude's JSON response
   */
  private parsePrioritiesFromResponse(
    responseText: string,
    conversations: LimitlessConversation[]
  ): Priority[] {
    try {
      // Extract JSON from response (handle markdown code blocks)
      const jsonMatch = responseText.match(/```json\s*([\s\S]*?)\s*```/) ||
                       responseText.match(/{[\s\S]*}/);

      if (!jsonMatch) {
        console.warn('[Claude] No JSON found in response');
        return [];
      }

      const jsonText = jsonMatch[1] || jsonMatch[0];
      const parsed = JSON.parse(jsonText);

      const priorities: Priority[] = parsed.priorities || [];

      // Enrich with source timestamps
      return priorities.map((p) => ({
        ...p,
        sourceTimestamp: this.findConversationTimestamp(p.source, conversations),
      }));

    } catch (error) {
      console.error('[Claude] Failed to parse priorities:', error);
      return [];
    }
  }

  /**
   * Find conversation timestamp from source description
   */
  private findConversationTimestamp(
    source: string,
    conversations: LimitlessConversation[]
  ): Date | undefined {
    // Try to match conversation by participants or title
    const conv = conversations.find(
      (c) =>
        c.title?.toLowerCase().includes(source.toLowerCase()) ||
        c.participants?.some((p) => source.toLowerCase().includes(p.toLowerCase()))
    );

    return conv ? new Date(conv.startTime) : undefined;
  }

  /**
   * Calculate overall confidence from priorities
   */
  private calculateOverallConfidence(priorities: Priority[]): number {
    if (priorities.length === 0) return 0;

    const avgConfidence =
      priorities.reduce((sum, p) => sum + p.confidence, 0) / priorities.length;

    return Math.round(avgConfidence * 100) / 100;
  }

  /**
   * Health check - verify API connectivity
   */
  async isConnected(): Promise<boolean> {
    try {
      await this.client.messages.create({
        model: this.model,
        max_tokens: 10,
        messages: [{ role: 'user', content: 'test' }],
      });
      return true;
    } catch {
      return false;
    }
  }
}
